Token.py

```diff
@@ -21,12 +22,160 @@ def __init__(
+        self.total_label = self.get_total_label()
+        self.date_label = self.get_date_label()
+        self.contains_digit = self.get_contains_digits()
+
     def __repr__(self):
         return self.text if self.text else str(self.token_structure)
 
     def __str__(self):
         return self.text if self.text else str(self.token_structure)
 
+    # returns true if has at least one digit, false otherwise
+    def get_contains_digits(self):
+        if self.text:
+            for c in self.text:
+                if c.isdigit():
+                    return True
+            return False
+
+    def get_date_label(self):
+        kw = ["date", "Date"]
+        if self.text:
+            for w in kw:
+                if w in self.text:
+                    return self.text
+
+    # tries to extract address from token
+    def get_address(self):
+        kw = [
+            "drive",
+            "dr",
+            "road",
+            "rd",
+            "lane",
+            "ln",
+            "ave",
+            "avenue",
+            "street",
+            "jalan",
+            "jln",
+            "boulevard",
+            "blvd",
+            "way",
+            "park",
+            "gate",
+            "crescent",
+            "cres",
+            "singapore",
+            "grove",
+            "grv"
+        ]
+        if self.text:
+            text_array = self.text.lower().split(" ")
+            for w in kw:
+                for t in text_array:
+                    if re.search("^" + w, t) and len(text_array) < 10:
+                        return self.text
+
+    # returns the text if "total" or some variant is contained in text and group is fewer than 5 words
+    def get_total_label(self):
+        kw = ["total", "Total"]
+        if self.text:
+            for w in kw:
+                if w in self.text and len(self.text.split(" ")) < 5:
+                    return self.text
+
+    # returns string for description of number, eg. account number, invoice number
+    def get_num_label(self):
+        kw = ["no.", "no:", "number", "num", "No.", "No:"]
+        if self.text:
+            text_array = self.text.split(" ")
+            for w in kw:
+                if w in text_array and text_array.index(w) > 0:
+                    return text_array[text_array.index(w) - 1]
+
+    # returns a dictionary of {cur: <prefix> , value: <dollar amt> }
+    # eg. {cur: $ , value: 5.00 }
+    def get_currency(self):
+        currencies = ["$", "¥", "dollar", "SGD", "USD", "US$", "SG$", "$SG", "$US"]
+        out = {}
+        for cur in currencies:
+            if self.text and cur in self.text:
+                out["cur"] = cur
+                start = self.text.index(cur) + len(cur)
+                for i in range(start, len(self.text)):
+                    char = self.text[i]
+                    if not char.isdigit() and not (char in [".", ","]):
+                        break
+                try:
```

Comment:
In principle, would be better to use an if else to catch whether there are digits after the currency sign, so that we know when a true exception occurs

Reply:
not sure how i would use for the conditional though. im just seeing if the float() function can return properly for this one

Reply:
Ah I see what you mean. Perhaps can use this:
```except ValueError as e:```
to make it specifically catch strings that cannot be converted to float